README.md
# curved_navigation_rail
[pub.dev ->](https://pub.dartlang.org/packages/curved_navigation_rail)

A Flutter package for easy implementation of curved navigation rail. Designed to be a ***direct drop-in replacement*** for [NavigationRail](https://api.flutter.dev/flutter/material/NavigationRail-class.html).

This is a fork of the [curved_navigation_bar](https://pub.dev/packages/curved_navigation_bar) package. Thank you to the developers of that package for inspiration. Now the same curved theme can consistently be applied in adaptive layouts!

![Gif](https://github.com/gangyistudios/curved_navigation_rail/blob/master/example/assets/example1.gif "NavigationRail drop-in example")

Retaining the same example as curved_navigation_bar:

![Gif](https://github.com/gangyistudios/curved_navigation_rail/blob/master/example/assets/example.gif "curved_navigation_rail example")

### Add dependency

```yaml
dependencies:
  curved_navigation_rail: ^2.0.1 #latest version
```

### Easy to use

- See minimal_example.dart in the examples

### Attributes

`destinations`: List of NavigationRailDestinations\
`selectedIndex`: index of NavigationRail, can be used to change current index or to set initial index\
`color`: Color of NavigationRail, default Colors.white\
`buttonBackgroundColor`: background color of floating button, default same as color attribute\
`backgroundColor`: Color of NavigationBar's background, defaults to the context theme's scaffoldBackgroundColor\
`onDestinationSelected`: Function handling taps on items, takes an int of the selectedIndex\
`animationCurve`: Curves interpolating button change animation, default Curves.easeOutCubic\
`animationDuration`: Duration of button change animation, default Duration(milliseconds: 600)\
`width`: Width of NavigationRail, min 0.0, max 75.0\
`letIndexChange`: Function which takes page index as argument and returns bool. If function returns false then page is not changed on button tap. It returns true by default\


### Examples 

1. minimal_example.dart - the bare minimum. 

2. change_page_programatically.dart - example of changing navigation rail item using a button on the page. 

3. navigation_rail_dropin.dart - direct replacement for NavigationRail. Just replace NavigationRail with CurvedNavigationRail!

4. navigation_rail_dropin_inverted.dart - extended example of navigation_rail_dropin.dart showing an inverted navigation rail, for right to left directionalities. (TO BE IMPLEMENTED). 


### Road map 

- Add support for Right-To-Left directionality, i.e. invert the CurvedNavigationRail when placed on the right of the screen.
- Add support for displaying text with the icon, and subsequently support for NavigationRailLabelType as a parameter input as used in NavigationRail. 
- Add parameter for height and alignment, to allow for greater customisation than just spaced evenly along the entire height. 
- Add support for adaptive layout, so that the rail extends to a drawer at a given width threshold. 


CHANGELOG.md
## [2.0.1]
- Tidy up package, code and example, no functional changes

## [2.0.0]
- Convert CurvedNavigationBar to CurvedNavigationRail widget 
- Intended to be a direct replacement for NavigationRail widget 
- No longer compatible with CurvedNavigationBar 

## [1.0.3]
- rollback 1.0.2 bug

## [1.0.2]
- update active menu item button CurvedNavigationBar widget change

## [1.0.1]
- make the example compatible with dart >= 2.12.0

## [1.0.0]
- merge null safety into master

## [0.4.0-nullsafety.0]
- null safety

## [0.3.7]
- Reduce package size

## [0.3.6]
- Add letIndexChange param to CurvedNavigationBar

## [0.3.5]
- Use clipBehavior instead of overflow, add example project

## [0.3.4]
- dartfmt format

## [0.3.3]
- By default center items

## [0.3.2]
- Detect tap on the whole NavButton

## [0.3.1]
- Replace InkWell by GestureDetector which wraps each of navigation buttons

## [0.3.0]
- Change page programmatically

## [0.2.20]
- TextDirection Right-To-Left and Left-To-Right auto support

## [0.2.11]
- remove debug print

## [0.2.1]
- height can be equal to 0.0

## [0.2.0]
- navigation bar index change available via index param, adjustable height via height param, minimum items: 1

## [0.1.26]
- Make Surface area for button clicks bigger

## [0.1.25]
- add buttonBackgroundColor param

## [0.1.24]
- readme grammar fix

## [0.1.23]
- fix initialIndex

## [0.1.22]
- make package usable on flutter stable version

## [0.1.2]
- update Readme dependency version

## [0.1.1]
- update Readme package usage and package link

## [0.1.0]
- update Readme

## [0.0.12]
- add Example

## [0.0.1]
- Initial version, 07.03.2018

test/curved_navigation_rail_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:curved_navigation_rail/curved_navigation_rail.dart';

void main() {
  test('', () {});
}


example/lib/main.dart
import 'package:curved_navigation_rail/curved_navigation_rail.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const App());
}

class App extends StatelessWidget {
  const App({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: CurvedNavigationRailExamplePage(),
    );
  }
}

class CurvedNavigationRailExamplePage extends StatefulWidget {
  const CurvedNavigationRailExamplePage({Key? key}) : super(key: key);

  @override
  State<CurvedNavigationRailExamplePage> createState() =>
      _CurvedNavigationRailExamplePageState();
}

class _CurvedNavigationRailExamplePageState
    extends State<CurvedNavigationRailExamplePage> {
  int _selectedIndex = 0;

  @override
  Widget build(BuildContext context) {
    final List<NavigationRailDestination> destinations = const [
      NavigationRailDestination(
          icon: Icon(Icons.person), label: Text('Profile')),
      NavigationRailDestination(icon: Icon(Icons.home), label: Text('Home')),
      NavigationRailDestination(
          icon: Icon(Icons.settings), label: Text('Settings')),
    ];

    return Scaffold(
        body: Row(
      children: [
        CurvedNavigationRail(
          destinations: destinations,
          onDestinationSelected: (index) =>
              setState(() => _selectedIndex = index),
        ),
        MyContent(value: _selectedIndex),
      ],
    ));
  }
}

class MyContent extends StatelessWidget {
  const MyContent({Key? key, required this.value}) : super(key: key);

  final int value;

  @override
  Widget build(BuildContext context) {
    return Expanded(child: Center(child: Text('This is Page $value')));
  }
}


example/lib/example1.dart
import 'package:flutter/material.dart';
import 'package:curved_navigation_rail/curved_navigation_rail.dart';

void main() => runApp(MaterialApp(home: CurvedNavigationRailExample()));

class CurvedNavigationRailExample extends StatefulWidget {
  @override
  _CurvedNavigationRailExampleState createState() =>
      _CurvedNavigationRailExampleState();
}

class _CurvedNavigationRailExampleState
    extends State<CurvedNavigationRailExample> {
  int _page = 0;
  GlobalKey<CurvedNavigationRailState> _bottomNavigationKey = GlobalKey();

  final List<Widget> destinationsOld = [
    Icon(Icons.add, size: 30),
    Icon(Icons.list, size: 30),
    Icon(Icons.compare_arrows, size: 30),
    Icon(Icons.call_split, size: 30),
    Icon(Icons.perm_identity, size: 30),
  ];

  final List<NavigationRailDestination> destinations = [
    NavigationRailDestination(
      icon: Icon(Icons.add),
      label: Text('First'),
    ),
    NavigationRailDestination(
      icon: Icon(Icons.list),
      label: Text('Second'),
    ),
    NavigationRailDestination(
      icon: Icon(Icons.compare_arrows),
      label: Text('Third'),
    ),
    NavigationRailDestination(
      icon: Icon(Icons.call_split),
      label: Text('Fourth'),
    ),
    NavigationRailDestination(
      icon: Icon(Icons.perm_identity),
      label: Text('Fifth'),
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        bottomNavigationBar: CurvedNavigationRail(
          key: _bottomNavigationKey,
          selectedIndex: 0,
          width: 60.0,
          destinations: destinations,
          color: Colors.white,
          buttonBackgroundColor: Colors.white,
          backgroundColor: Colors.blue,
          animationCurve: Curves.easeInOut,
          animationDuration: Duration(milliseconds: 600),
          onDestinationSelected: (index) {
            setState(() {
              _page = index;
            });
          },
          letIndexChange: (index) => true,
        ),
        body: Container(
          color: Colors.blue,
          child: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Text(_page.toString(), textScaleFactor: 10.0),
                ElevatedButton(
                  child: Text('Go To Page of index 1'),
                  onPressed: () {
                    final CurvedNavigationRailState? navBarState =
                        _bottomNavigationKey.currentState;
                    navBarState?.setPage(1);
                  },
                )
              ],
            ),
          ),
        ));
  }
}


example/lib/example2.dart
import 'package:flutter/material.dart';
import 'package:curved_navigation_rail/curved_navigation_rail.dart';

/// Drop-in replacement for NavigationRail using the official example at
/// https://api.flutter.dev/flutter/material/NavigationRail-class.html
///
/// ** NOTE ** labelType parameter is currently not supported.
void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: _title,
      home: MyStatefulWidget(),
    );
  }
}

class MyStatefulWidget extends StatefulWidget {
  const MyStatefulWidget({Key? key}) : super(key: key);

  @override
  State<MyStatefulWidget> createState() => _MyStatefulWidgetState();
}

class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  int _selectedIndex = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Row(
        children: <Widget>[
          CurvedNavigationRail(
            selectedIndex: _selectedIndex,
            onDestinationSelected: (int index) {
              setState(() {
                _selectedIndex = index;
              });
            },
            destinations: const <NavigationRailDestination>[
              NavigationRailDestination(
                icon: Icon(Icons.favorite_border),
                selectedIcon: Icon(Icons.favorite),
                label: Text('First'),
              ),
              NavigationRailDestination(
                icon: Icon(Icons.bookmark_border),
                selectedIcon: Icon(Icons.book),
                label: Text('Second'),
              ),
              NavigationRailDestination(
                icon: Icon(Icons.star_border),
                selectedIcon: Icon(Icons.star),
                label: Text('Third'),
              ),
            ],
          ),
          // This is the main content.
          Expanded(
            child: Center(
              child: Text('selectedIndex: $_selectedIndex'),
            ),
          )
        ],
      ),
    );
  }
}


lib/curved_navigation_rail.dart
import 'package:flutter/material.dart';
import 'src/nav_button.dart';
import 'src/nav_custom_painter.dart';

typedef _LetIndexPage = bool Function(int value);

class CurvedNavigationRail extends StatefulWidget {
  final List<NavigationRailDestination> destinations;
  final int selectedIndex;
  final Color color;
  final Color? buttonBackgroundColor;
  final Color? backgroundColor;
  final ValueChanged<int>? onDestinationSelected;
  final _LetIndexPage letIndexChange;
  final Curve animationCurve;
  final Duration animationDuration;
  final double width;

  CurvedNavigationRail({
    Key? key,
    required this.destinations,
    this.selectedIndex = 0,
    this.color = Colors.white,
    this.buttonBackgroundColor,
    this.backgroundColor,
    this.onDestinationSelected,
    _LetIndexPage? letIndexChange,
    this.animationCurve = Curves.easeOut,
    this.animationDuration = const Duration(milliseconds: 600),
    this.width = 75.0,
  })  : letIndexChange = letIndexChange ?? ((_) => true),
        assert(destinations.length >= 1),
        assert(0 <= selectedIndex && selectedIndex < destinations.length),
        assert(0 <= width && width <= 75.0),
        super(key: key);

  @override
  CurvedNavigationRailState createState() => CurvedNavigationRailState();
}

class CurvedNavigationRailState extends State<CurvedNavigationRail>
    with SingleTickerProviderStateMixin {
  late double _startingPos;
  int _endingIndex = 0;
  late double _pos;
  double _buttonHide = 0;
  late Widget _icon;
  late AnimationController _animationController;
  late int _length;

  @override
  void initState() {
    super.initState();
    _icon = widget.destinations[widget.selectedIndex].icon;
    _length = widget.destinations.length;
    _pos = widget.selectedIndex / _length;
    _startingPos = widget.selectedIndex / _length;
    _animationController = AnimationController(vsync: this, value: _pos);
    _animationController.addListener(() {
      setState(() {
        _pos = _animationController.value;
        final endingPos = _endingIndex / widget.destinations.length;
        final middle = (endingPos + _startingPos) / 2;
        if ((endingPos - _pos).abs() < (_startingPos - _pos).abs()) {
          _icon = widget.destinations[_endingIndex].icon;
        }
        _buttonHide =
            (1 - ((middle - _pos) / (_startingPos - middle)).abs()).abs();
      });
    });
  }

  @override
  void didUpdateWidget(CurvedNavigationRail oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.selectedIndex != widget.selectedIndex) {
      final newPosition = widget.selectedIndex / _length;
      _startingPos = _pos;
      _endingIndex = widget.selectedIndex;
      _animationController.animateTo(newPosition,
          duration: widget.animationDuration, curve: widget.animationCurve);
    }
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    return Row(children: [
      Container(
        color:
            widget.backgroundColor ?? Theme.of(context).scaffoldBackgroundColor,
        width: widget.width,
        child: Stack(
          clipBehavior: Clip.none,
          alignment: Alignment.centerLeft,
          children: [
            SelectedIcon(
              widget: widget,
              pos: _pos,
              size: size,
              length: _length,
              buttonHide: _buttonHide,
              icon: _icon,
            ),
            CurvedRail(widget: widget, pos: _pos, length: _length),
            UnselectedIcons(
                widget: widget, pos: _pos, length: _length, onTap: _buttonTap),
          ],
        ),
      ),
    ]);
  }

  void setPage(int index) {
    _buttonTap(index);
  }

  void _buttonTap(int index) {
    if (!widget.letIndexChange(index)) {
      return;
    }
    if (widget.onDestinationSelected != null) {
      widget.onDestinationSelected!(index);
    }
    final newPosition = index / _length;
    setState(() {
      _startingPos = _pos;
      _endingIndex = index;
      _animationController.animateTo(newPosition,
          duration: widget.animationDuration, curve: widget.animationCurve);
    });
  }
}

class UnselectedIcons extends StatelessWidget {
  const UnselectedIcons({
    Key? key,
    required this.widget,
    required this.pos,
    required this.length,
    required this.onTap,
  }) : super(key: key);

  final CurvedNavigationRail widget;
  final double pos;
  final int length;
  final void Function(int) onTap;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: 0 - (75.0 - widget.width),
      top: 0,
      bottom: 0,
      child: SizedBox(
          width: 100.0,
          child: Column(
              children: widget.destinations.map((item) {
            return NavButton(
              onTap: onTap,
              position: pos,
              length: length,
              index: widget.destinations.indexOf(item),
              child: Center(child: item.icon),
            );
          }).toList())),
    );
  }
}

class CurvedRail extends StatelessWidget {
  const CurvedRail({
    Key? key,
    required this.widget,
    required double pos,
    required int length,
  })  : _pos = pos,
        _length = length,
        super(key: key);

  final CurvedNavigationRail widget;
  final double _pos;
  final int _length;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: 0 - (75.0 - widget.width),
      top: 0,
      bottom: 0,
      child: CustomPaint(
        painter: NavCustomPainter(
            _pos, _length, widget.color, Directionality.of(context)),
        child: Container(
          width: 75.0,
        ),
      ),
    );
  }
}

class SelectedIcon extends StatelessWidget {
  const SelectedIcon({
    Key? key,
    required this.widget,
    required double pos,
    required this.size,
    required int length,
    required double buttonHide,
    required Widget icon,
  })  : _pos = pos,
        _length = length,
        _buttonHide = buttonHide,
        _icon = icon,
        super(key: key);

  final CurvedNavigationRail widget;
  final double _pos;
  final Size size;
  final int _length;
  final double _buttonHide;
  final Widget _icon;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      top: _pos * size.height,
      left: -40 - (75.0 - widget.width),
      height: size.height / _length,
      child: SelectedIconCircle(
        buttonHide: _buttonHide,
        widget: widget,
        icon: _icon,
      ),
    );
  }
}

class SelectedIconCircle extends StatelessWidget {
  const SelectedIconCircle({
    Key? key,
    required double buttonHide,
    required this.widget,
    required Widget icon,
  })  : _buttonHide = buttonHide,
        _icon = icon,
        super(key: key);

  final double _buttonHide;
  final CurvedNavigationRail widget;
  final Widget _icon;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Transform.translate(
        offset: Offset(
          (1 - _buttonHide) * 80,
          0,
        ),
        child: Material(
          color: widget.buttonBackgroundColor ?? widget.color,
          type: MaterialType.circle,
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: _icon,
          ),
        ),
      ),
    );
  }
}


lib/src/nav_custom_painter.dart
import 'dart:ui';

import 'package:flutter/material.dart';

class NavCustomPainter extends CustomPainter {
  late double loc;
  late double s;
  Color color;
  TextDirection textDirection;

  NavCustomPainter(
      double startingLoc, int itemsLength, this.color, this.textDirection) {
    final span = 1.0 / itemsLength;
    s = 0.2;
    double l = startingLoc + (span - s) / 2;
    loc = textDirection == TextDirection.rtl ? 0.8 - l : l;
  }

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = color
      ..style = PaintingStyle.fill;

    final double w = size.width;
    final double h = size.height;

    List<Point> p = [
      Point(w, 0),
      Point(w, (loc - 0.1) * h),
      Point(w * 0.95, (loc + s * 0.20) * h),
      Point(w * 0.4, loc * h),
      Point(w * 0.4, (loc + s * 0.5) * h),
      Point(w * 0.4, (loc + s) * h),
      Point(w * 0.95, (loc + s - s * 0.2) * h),
      Point(w, (loc + s + 0.1) * h),
      Point(w, h),
      Point(0, h),
      Point(0, 0),
    ];

    final path = Path()
      ..moveTo(p[0].x, p[0].y)
      ..lineTo(p[1].x, p[1].y)
      ..cubicTo(p[2].x, p[2].y, p[3].x, p[3].y, p[4].x, p[4].y)
      ..cubicTo(p[5].x, p[5].y, p[6].x, p[6].y, p[7].x, p[7].y)
      ..lineTo(p[8].x, p[8].y)
      ..lineTo(p[9].x, p[9].y)
      ..lineTo(p[10].x, p[10].y)
      ..close();
    canvas.drawPath(path, paint);

    // visualisePoints(canvas, p);
  }

  void visualisePoints(Canvas canvas, List<Point> points) {
    var paint = Paint()
      ..color = Colors.brown
      ..strokeWidth = 20;
    canvas.drawPoints(
        PointMode.points, points.map((e) => Offset(e.x, e.y)).toList(), paint);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return this != oldDelegate;
  }
}

class Point {
  final double x;
  final double y;

  Point(
    this.x,
    this.y,
  );

  @override
  String toString() {
    return ('($x, $y)');
  }
}


example/lib/change_page_programatically.dart
import 'package:flutter/material.dart';
import 'package:curved_navigation_rail/curved_navigation_rail.dart';

void main() => runApp(
      const MaterialApp(
        home: CurvedNavigationRailExample(),
      ),
    );

class CurvedNavigationRailExample extends StatefulWidget {
  const CurvedNavigationRailExample({Key? key}) : super(key: key);

  @override
  _CurvedNavigationRailExampleState createState() =>
      _CurvedNavigationRailExampleState();
}

class _CurvedNavigationRailExampleState
    extends State<CurvedNavigationRailExample> {
  final GlobalKey<CurvedNavigationRailState> _bottomNavigationKey = GlobalKey();
  int _page = 0;

  final List<NavigationRailDestination> destinations = const [
    NavigationRailDestination(
      icon: Icon(Icons.add),
      label: Text('First'),
    ),
    NavigationRailDestination(
      icon: Icon(Icons.list),
      label: Text('Second'),
    ),
    NavigationRailDestination(
      icon: Icon(Icons.compare_arrows),
      label: Text('Third'),
    ),
    NavigationRailDestination(
      icon: Icon(Icons.call_split),
      label: Text('Fourth'),
    ),
    NavigationRailDestination(
      icon: Icon(Icons.perm_identity),
      label: Text('Fifth'),
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.blue,
      body: Row(
        children: [
          CurvedNavigationRail(
            key: _bottomNavigationKey,
            selectedIndex: 0,
            width: 60.0,
            destinations: destinations,
            color: Colors.white,
            buttonBackgroundColor: Colors.white,
            backgroundColor: Colors.blue,
            animationCurve: Curves.easeInOut,
            animationDuration: const Duration(milliseconds: 600),
            onDestinationSelected: (index) {
              setState(() {
                _page = index;
              });
            },
            letIndexChange: (index) => true,
          ),
          Expanded(
            child: Center(
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Text(_page.toString(), textScaleFactor: 10.0),
                  ElevatedButton(
                    child: const Text('Go To Page of index 1'),
                    onPressed: () {
                      final CurvedNavigationRailState? navBarState =
                          _bottomNavigationKey.currentState;
                      navBarState?.setPage(1);
                    },
                  )
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}


example/lib/navigation_rail_dropin_inverted.dart
import 'package:flutter/material.dart';
import 'package:curved_navigation_rail/curved_navigation_rail.dart';

/// Drop-in replacement for NavigationRail using the official example at
/// https://api.flutter.dev/flutter/material/NavigationRail-class.html
///
/// ** NOTE ** labelType parameter is currently not supported.
///
/// This is the inverted version which is for right to left directionalities.
/// TODO: Add directionality support and update this example.
void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: _title,
      home: MyStatefulWidget(),
    );
  }
}

class MyStatefulWidget extends StatefulWidget {
  const MyStatefulWidget({Key? key}) : super(key: key);

  @override
  State<MyStatefulWidget> createState() => _MyStatefulWidgetState();
}

class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  int _selectedIndex = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Row(
        children: <Widget>[
          CurvedNavigationRail(
            selectedIndex: _selectedIndex,
            onDestinationSelected: (int index) {
              setState(() {
                _selectedIndex = index;
              });
            },
            destinations: const <NavigationRailDestination>[
              NavigationRailDestination(
                icon: Icon(Icons.favorite_border),
                selectedIcon: Icon(Icons.favorite),
                label: Text('First'),
              ),
              NavigationRailDestination(
                icon: Icon(Icons.bookmark_border),
                selectedIcon: Icon(Icons.book),
                label: Text('Second'),
              ),
              NavigationRailDestination(
                icon: Icon(Icons.star_border),
                selectedIcon: Icon(Icons.star),
                label: Text('Third'),
              ),
            ],
          ),
          // This is the main content.
          Expanded(
            child: Center(
              child: Text('selectedIndex: $_selectedIndex'),
            ),
          )
        ],
      ),
    );
  }
}


lib/curved_navigation_bar.dart
import 'package:flutter/material.dart';
import 'package:meta/meta.dart';
import 'src/nav_button.dart';
import 'src/nav_custom_painter.dart';

typedef _LetIndexPage = bool Function(int value);

class CurvedNavigationBar extends StatefulWidget {
  final List<Widget> items;
  final int index;
  final Color color;
  final Color? buttonBackgroundColor;
  final Color backgroundColor;
  final ValueChanged<int>? onTap;
  final _LetIndexPage letIndexChange;
  final Curve animationCurve;
  final Duration animationDuration;
  final double height;

  CurvedNavigationBar({
    Key? key,
    required this.items,
    this.index = 0,
    this.color = Colors.white,
    this.buttonBackgroundColor,
    this.backgroundColor = Colors.blueAccent,
    this.onTap,
    _LetIndexPage? letIndexChange,
    this.animationCurve = Curves.easeOut,
    this.animationDuration = const Duration(milliseconds: 600),
    this.height = 75.0,
  })  : letIndexChange = letIndexChange ?? ((_) => true),
        assert(items != null),
        assert(items.length >= 1),
        assert(0 <= index && index < items.length),
        assert(0 <= height && height <= 75.0),
        super(key: key);

  @override
  CurvedNavigationBarState createState() => CurvedNavigationBarState();
}

class CurvedNavigationBarState extends State<CurvedNavigationBar>
    with SingleTickerProviderStateMixin {
  late double _startingPos;
  int _endingIndex = 0;
  late double _pos;
  double _buttonHide = 0;
  late Widget _icon;
  late AnimationController _animationController;
  late int _length;

  @override
  void initState() {
    super.initState();
    _icon = widget.items[widget.index];
    _length = widget.items.length;
    _pos = widget.index / _length;
    _startingPos = widget.index / _length;
    _animationController = AnimationController(vsync: this, value: _pos);
    _animationController.addListener(() {
      setState(() {
        _pos = _animationController.value;
        final endingPos = _endingIndex / widget.items.length;
        final middle = (endingPos + _startingPos) / 2;
        if ((endingPos - _pos).abs() < (_startingPos - _pos).abs()) {
          _icon = widget.items[_endingIndex];
        }
        _buttonHide =
            (1 - ((middle - _pos) / (_startingPos - middle)).abs()).abs();
      });
    });
  }

  @override
  void didUpdateWidget(CurvedNavigationBar oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.index != widget.index) {
      final newPosition = widget.index / _length;
      _startingPos = _pos;
      _endingIndex = widget.index;
      _animationController.animateTo(newPosition,
          duration: widget.animationDuration, curve: widget.animationCurve);
    }
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    return Row(children: [
      Container(
        color: widget.backgroundColor,
        // TODO: change this to a width variable.
        width: widget.height,
        // height: 800,
        child: Stack(
          clipBehavior: Clip.none,
          alignment: Alignment.centerLeft,
          children: [
            Scaffold(body: Center(child: Text('hello'))),
            Positioned1a(
              widget: widget,
              pos: _pos,
              size: size,
              length: _length,
              buttonHide: _buttonHide,
              icon: _icon,
            ),
            Positioned2a(widget: widget, pos: _pos, length: _length),
            Positioned(
              // left: 0,
              // right: 0,
              // bottom: 0 - (75.0 - widget.height),
              // TODO: Change to widget.width
              left: 0 - (75.0 - widget.height),
              top: 0,
              bottom: 0,
              child: SizedBox(
                  width: 100.0,
                  child: Column(
                      children: widget.items.map((item) {
                    return NavButton(
                      onTap: _buttonTap,
                      position: _pos,
                      length: _length,
                      index: widget.items.indexOf(item),
                      child: Center(child: item),
                    );
                  }).toList())),
            ),
          ],
        ),
      ),
    ]);

    return Container(
      color: widget.backgroundColor,
      height: widget.height,
      child: Stack(
        clipBehavior: Clip.none,
        alignment: Alignment.bottomCenter,
        children: <Widget>[
          Positioned1(
              widget: widget,
              pos: _pos,
              size: size,
              length: _length,
              buttonHide: _buttonHide,
              icon: _icon),
          Positioned2(widget: widget, pos: _pos, length: _length),
          Positioned(
            left: 0,
            right: 0,
            bottom: 0 - (75.0 - widget.height),
            child: SizedBox(
                height: 100.0,
                child: Row(
                    children: widget.items.map((item) {
                  return NavButton(
                    onTap: _buttonTap,
                    position: _pos,
                    length: _length,
                    index: widget.items.indexOf(item),
                    child: Center(child: item),
                  );
                }).toList())),
          ),
        ],
      ),
    );
  }

  void setPage(int index) {
    _buttonTap(index);
  }

  void _buttonTap(int index) {
    if (!widget.letIndexChange(index)) {
      return;
    }
    if (widget.onTap != null) {
      widget.onTap!(index);
    }
    final newPosition = index / _length;
    setState(() {
      _startingPos = _pos;
      _endingIndex = index;
      _animationController.animateTo(newPosition,
          duration: widget.animationDuration, curve: widget.animationCurve);
    });
  }
}

class Positioned2a extends StatelessWidget {
  const Positioned2a({
    Key? key,
    required this.widget,
    required double pos,
    required int length,
  })  : _pos = pos,
        _length = length,
        super(key: key);

  final CurvedNavigationBar widget;
  final double _pos;
  final int _length;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      // left: 0,
      // right: 0,
      // bottom: 0 - (75.0 - widget.height),
      // NEW
      // TODO: Change to widget.width
      left: 0 - (75.0 - widget.height),
      top: 0,
      bottom: 0,
      child: CustomPaint(
        painter: NavCustomPainter(
            _pos, _length, widget.color, Directionality.of(context)),
        child: Container(
          width: 75.0,
        ),
      ),
    );
  }
}

class Positioned2 extends StatelessWidget {
  const Positioned2({
    Key? key,
    required this.widget,
    required double pos,
    required int length,
  })  : _pos = pos,
        _length = length,
        super(key: key);

  final CurvedNavigationBar widget;
  final double _pos;
  final int _length;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: 0,
      right: 0,
      bottom: 0 - (75.0 - widget.height),
      child: CustomPaint(
        painter: NavCustomPainter(
            _pos, _length, widget.color, Directionality.of(context)),
        child: Container(
          height: 75.0,
        ),
      ),
    );
  }
}

class Positioned1a extends StatelessWidget {
  const Positioned1a({
    Key? key,
    required this.widget,
    required double pos,
    required this.size,
    required int length,
    required double buttonHide,
    required Widget icon,
  })  : _pos = pos,
        _length = length,
        _buttonHide = buttonHide,
        _icon = icon,
        super(key: key);

  final CurvedNavigationBar widget;
  final double _pos;
  final Size size;
  final int _length;
  final double _buttonHide;
  final Widget _icon;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      top: _pos * size.height,
      left: -40 - (75.0 - widget.height),
      height: size.height / _length,
      child: Positioned1aCenter(
        buttonHide: _buttonHide,
        widget: widget,
        icon: _icon,
      ),
    );
  }
}

class Positioned1aCenter extends StatelessWidget {
  const Positioned1aCenter({
    Key? key,
    required double buttonHide,
    required this.widget,
    required Widget icon,
  })  : _buttonHide = buttonHide,
        _icon = icon,
        super(key: key);

  final double _buttonHide;
  final CurvedNavigationBar widget;
  final Widget _icon;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Transform.translate(
        offset: Offset(
          (1 - _buttonHide) * 80,
          0,
        ),
        child: Material(
          color: widget.buttonBackgroundColor ?? widget.color,
          type: MaterialType.circle,
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: _icon,
          ),
        ),
      ),
    );
  }
}

class Positioned1 extends StatelessWidget {
  const Positioned1({
    Key? key,
    required this.widget,
    required double pos,
    required this.size,
    required int length,
    required double buttonHide,
    required Widget icon,
  })  : _pos = pos,
        _length = length,
        _buttonHide = buttonHide,
        _icon = icon,
        super(key: key);

  final CurvedNavigationBar widget;
  final double _pos;
  final Size size;
  final int _length;
  final double _buttonHide;
  final Widget _icon;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      bottom: -40 - (75.0 - widget.height),
      left: Directionality.of(context) == TextDirection.rtl
          ? null
          : _pos * size.width,
      right: Directionality.of(context) == TextDirection.rtl
          ? _pos * size.width
          : null,
      width: size.width / _length,
      child: Center(
        child: Transform.translate(
          offset: Offset(
            0,
            -(1 - _buttonHide) * 80,
          ),
          child: Material(
            color: widget.buttonBackgroundColor ?? widget.color,
            type: MaterialType.circle,
            child: Padding(
              padding: const EdgeInsets.all(8.0),
              child: _icon,
            ),
          ),
        ),
      ),
    );
  }
}
